### PR TITLE
Observe directly-opened desktop-app sessions (persistent WS observer + on-disk reconcile)

### LINKED ISSUE #119: Observe directly-opened desktop-app sessions (persistent WS observer + on-disk poll) + link back
Part of #101. **The real pain** and the hard part: sessions the user opens **directly in the Copilot desktop app** (not delegated from Projects) are never tracked, so they can't be shown or assigned. This is spike-flavored - it rides the app's unofficial local protocol - so timebox it and be honest about fragility.

## Goal
When a Copilot session is opened directly in the desktop app in a repo Projects knows about (e.g. the usersd repo), Projects observes it, maps it to a project by repo, shows it in that project (the Copilot tab from #117), and can link back to it.

## Good news (de-risks the "can we link back?" question)
The app supports deep links to a session: `github-app://sessions/<id>` via `buildAppSessionDeepLink` (already used for delegated sessions in `src/main/index.ts`). So as long as we capture a session **id**, linking back already works. No new app cooperation needed for the link.

## Two complementary mechanisms (the WS alone isn't enough)
1. **Persistent WS observer.** Today the WS client is one-shot (connect -> create -> send -> disconnect) and only tracks sessions Projects launched. The protocol comment says the app emits `session_event` **"for ALL sessions."** So keep a persistent listener (isolated in `src/main/agent/copilot-app/`) that ingests session frames for sessions Projects didn't create. Handle: token rotates per app launch (re-read `~/.copilot/run/ws.{port,token}` on reconnect), reconnect on app relaunch, and the app being closed (degrade quietly).
   - **Open question the spike must answer:** does `session_event` (for externally-created sessions) carry the `cwd`? `session_created` does, but we only see that for sessions we create. If `session_event` lacks `cwd`, we get a session **id** but not the repo - which is exactly why mechanism 2 is needed.
2. **On-disk session-state poll.** Reconcile against the desktop app's own session store under `~/.copilot/` (session-state dirs / history). This catches sessions opened before we connected and supplies the `cwd` (hence repo) when the WS doesn't. Read-only; treat the layout as unofficial and fail closed if it changes.

## Mapping + storage
- Map session `cwd` -> repo (reuse the existing repos-root <-> owner/repo path logic in `src/main/agent/copilot-app/cwd.ts` / `resolveLocalCwd`) -> project via `repo_rules` / `resolveProjectId`.
- Upsert observed sessions into the copilot-app session store (`copilot_app_sessions`), tagged as observed (not Projects-launched), with `pinned_project_id` sticky semantics. They flow into the Copilot tab (#117) and the digest like launched sessions.
- Link back with the existing `github-app://sessions/<id>` deep link.

## Honesty / guardrails
- Unofficial protocol behind ONE adapter; an app update can only break this seam. Never log/leak the WS token. Read-only w.r.t. app files.
- Only works while the desktop app is running; degrade gracefully when it isn't.
- If the spike finds `session_event` carries no `cwd` AND the on-disk store can't be reliably read, document that and fall back to the MCP-skill self-registration idea (a `register_session(repo)` tool the SKILL.md tells Copilot to call) as a separate follow-up - do not silently ship something that half-works.

## Constraints
TS strict, no `any`, named exports, async/await, all I/O off the render thread, typed IPC. Vitest for the frame ingestion, cwd->repo->project mapping, reconnect/token-rotation handling, and the on-disk reconciler (against fixture dirs). Honest gap allowed (state in PR): a full live end-to-end against the real desktop app opening a session isn't automated.

## Verify
Drive a fake WS server emitting `session_event` for an unknown session + a fixture on-disk store, and assert Projects ingests it, maps cwd->repo->project, upserts an observed session, and exposes the deep link. Document what the real desktop app actually emits (the `cwd`-on-`session_event` question) from a manual check.

### PR DESCRIPTION
**Pull request by robert-crandall**
Closes #119. Part of #101 (epic #98).

This tracks Copilot sessions the user opens **directly in the desktop app** (not delegated from Projects) - the real pain from #101. It maps them to a project by repo, stores them as `observed` (distinct from Projects-`launched`), and lets us link back via the existing `github-app://sessions/<id>` deep link. It's spike-flavored: it rides the desktop app's unofficial local protocol, so it's built to degrade quietly and fail closed, all contained behind the one adapter dir `src/main/agent/copilot-app/`.

## The key open question, answered with live evidence

**Does the WS `session_event` (for externally-created sessions) carry the `cwd`? No.** I ran a read-only observer probe against the running app (token redacted, only listened, never sent a session-mutating frame). `session_event` is:

```
{ session_id, type, event: { type, data, id, parentId, timestamp, ephemeral } }
```

`event.type` values seen: `assistant.turn_start` / `turn_end`, `tool.execution_partial_result` / `_complete`, `session.usage_info`, `session.canvas.opened`, `session.background_tasks_changed`. A recursive key scan found **no `cwd` / `repository` / `git_root` anywhere**. So the WS tells us a session exists and is *live*, but not *where*. That's exactly why the on-disk reconciler is required, and it's the design here.

The repo/cwd comes from the app's own on-disk store: `~/.copilot/session-state/<id>/workspace.yaml` carries `cwd`, `repository` (owner/repo), and `name`. The directory basename equals the session id equals the WS `session_id`.

## Two mechanisms (both behind `src/main/agent/copilot-app/`)

1. **Persistent WS observer** (`observer.ts`) - a single long-lived listener, distinct from the one-shot delegate client. It treats `session_event` as a *liveness trigger*: an event for a session id triggers a targeted on-disk reconcile of that id. It re-reads the rotating token on every reconnect (never logs/throws it), resets backoff only on `server_hello` (a stable connection), uses capped exponential reconnect backoff, and a generation token so `stop()` turns every scheduled callback into a no-op. A short miss-retry schedule handles the `workspace.yaml` flush race, and a periodic scan is the backstop.
2. **On-disk reconciler** (`reconcile.ts`) - read-only. Parses `workspace.yaml` (defensively, no YAML dep) as the source of truth for repo, maps repo -> project via `resolveProjectId`, and upserts. Bounded by recency + a cap so the ~3,300 historical session dirs don't churn; `reconcileOne` is *not* recency-bounded so an old-but-active session flagged by the WS is still ingested. The dir basename is canonical identity; a `workspace.yaml.id` that disagrees is skipped. The asserted `repository` is validated to strict `owner/repo` and used **only** for project mapping - never for command execution or file access.

**Scope decision (honest):** I only store observed sessions whose repo resolves to a live project (i.e. "a repo Projects knows about"). Sessions in unmapped repos are skipped rather than piling thousands of unrelated app sessions into the store. They become observable if a mapping is added later while they're still in the recency window. A bounded unresolved-session cache for #118-style adoption is intentionally out of scope here.

## Storage / IPC (clean seam for #117/#118)

- Migration `021` adds `origin` (`launched` | `observed`, CHECK) and `pinned_project_id` (FK, `ON DELETE SET NULL`) to `copilot_app_sessions`.
- `upsertObservedSession` never downgrades a `launched` row to `observed`, and keeps a sticky `pinned_project_id` (pin wins while its project is live, else the freshly-resolved project) - mirroring the cloud `copilot_sessions` upsert.
- `assignAppSession` for manual sticky assignment; `deleteProject` now detaches app sessions the same way it detaches cloud sessions.
- New `copilot:project-app-sessions` IPC returns all of a project's app sessions (both origins), status-refreshed read-only. Deep link reuses the existing `copilot:open-app-session`.
- A visible **Settings toggle** ("Observe desktop-app sessions", default on) is a real kill switch that tears the whole pipeline down.

## Guardrails
- Read-only w.r.t. all `~/.copilot/` files; never writes them; never logs the WS token.
- Unofficial layout/protocol behind one adapter; fails closed on shape drift; token-safe diagnostic counters so a silent break is still debuggable.
- Only works while the app is running; degrades quietly otherwise. TS strict, no `any`, named exports, all I/O off the render thread, typed IPC.

## How I verified

- **`bun run test` green: 723 tests, typecheck + eslint clean.**
- **Unit/integration:**
  - `workspace-yaml.test.ts` - parser edge cases: quoted/doubled-quote scalars, block-scalar folding (`name: |-`), indented/nested keys ignored, comments, first-occurrence-wins, unterminated quote, strict `owner/repo` validation, session-id shape.
  - `reconcile.integration.test.ts` (real store + in-memory DB + fixture fs) - cwd->repo->project->observed upsert; skips no-repo / unresolved / already-launched (no downgrade) / id-mismatch; full scan is recency-bounded but `reconcileOne` still ingests an old session; sticky pin doesn't drift; no-op re-reconcile reports no change.
  - `observer.integration.test.ts` (real fake WS server) - `server_hello`-driven full reconcile; a `session_event` for a session not in the on-disk listing still ingests via targeted reconcile; reconnect + endpoint re-read after the app drops the socket; the `workspace.yaml` flush-race retry; a full reconcile doesn't drop a queued targeted id; `stop()` halts ingestion and reconnection.
  - store tests - observed upsert, no origin downgrade, sticky pin (incl. cleared when the pinned project is soft-deleted), `assignAppSession`, `deleteProject` detaches app sessions, still invisible to the github-session readers.
- **Live characterization on the running app** (this machine):
  - The `session_event`-has-no-cwd finding above.
  - The real reconciler parser run against the actual `~/.copilot/session-state`: 1,676 recent dirs scanned, 1,649 with a `cwd`, 1,464 with a valid `repository`, 0 id-mismatches, 43 correctly mapped to `robert-crandall/gh-notifier` (with correct cwds + titles).
  - The full WS->disk handoff against the live app: a real `session_event` for a session resolved through its on-disk `workspace.yaml` to `robert-crandall-org/librarian` with its real title.

## Honest gaps
- A fully automated end-to-end against the real desktop app opening a *brand-new* session is not automated (the WS + on-disk halves are each characterized live above, and the wiring is unit-tested against a fake server).
- The `workspace.yaml` layout is unofficial; if the app changes it, this one adapter fails closed (empty diagnostics counters) rather than misbehaving.
- The MCP-skill `register_session(repo)` self-registration idea from #119 is **not** needed here (session_event lacks cwd, but the on-disk `workspace.yaml` reliably supplies cwd+repo) - it stays a documented future option.

### CODE REVIEW

File: src/main/agent/copilot-app/workspace-yaml.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    if (!closed) return null // no closing quote → malformed
+    return out.length > 0 ? out : null
+  }
+
+  if (s.startsWith('"')) {
+    const end = s.indexOf('"', 1)
+    if (end === -1) return null
+    const body = s.slice(1, end).replace(/\\"/g, '"').replace(/\\\\/g, '\\')
+    return body.length > 0 ? body : null
+  }
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/workspace-yaml.ts:119]: Good catch - fixed in 2963af7. The double-quote branch now scans char-by-char honoring `\"` and `\\` escapes (mirroring the single-quote branch) instead of `indexOf`, so an escaped quote no longer truncates the value. Added tests for escaped quotes and backslashes.

File: src/main/agent/copilot-app/store.ts [hunk 2]
```diff
@@ -61,9 +65,9 @@ export function insertAppSession(input: InsertAppSessionInput): CopilotAppSessio
 
   db.prepare(`
     INSERT INTO copilot_app_sessions
-      (id, project_id, cwd, title, status, repo_owner, repo_name)
+      (id, project_id, cwd, title, status, repo_owner, repo_name, origin)
     VALUES
-      (?, ?, ?, ?, 'in_progress', ?, ?)
+      (?, ?, ?, ?, 'in_progress', ?, ?, 'launched')
     ON CONFLICT(id) DO UPDATE SET
       cwd        = excluded.cwd,
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/store.ts:75]: Fixed in 921f268. `insertAppSession`'s ON CONFLICT now sets `origin = 'launched'`, so an existing observed row gets upgraded to launched (authoritative provenance - Projects definitively created it). Added a regression test.

File: src/main/agent/copilot-app/store.ts [hunk 3]
```diff
@@ -76,10 +80,86 @@ export function insertAppSession(input: InsertAppSessionInput): CopilotAppSessio
...
+           WHERE p.id = copilot_app_sessions.pinned_project_id AND p.deleted_at IS NULL
+         )
+        THEN copilot_app_sessions.pinned_project_id
+        ELSE excluded.project_id
+      END,
+      cwd        = excluded.cwd,
+      title      = excluded.title,
+      repo_owner = excluded.repo_owner,
+      repo_name  = excluded.repo_name,
+      updated_at = datetime('now')
```
[the remark this answers is not in the document]
Comment (robert-crandall) [reply src/main/agent/copilot-app/store.ts:166]: Fixed in 921f268. `upsertObservedSession` now reads the existing row and skips the write entirely when it would come out identical (same repo/cwd/title and the same effective project/pin, mirroring the SQL's sticky-pin resolution), so a periodic reconcile no longer bumps `updated_at` or reorders the list. Added a regression test asserting an unchanged re-reconcile leaves `updated_at` untouched while a real change still writes.
